util/pyclient/bft_client.py

```diff
@@ -241,3 +247,257 @@ def get_rsi_replies(self):
+        return os.path.join(self.config.certs_path, str(replica_id), cert_type, cert_type + ".cert")
+
+    async def _establish_ssl_stream(self, dest_replica):
+        """
+        Try every 250ms to connect to dest_replica. We relay on the caller context that will trigger cancalleation at 
+        some point in time. Break the loop on success and keep the ssl_stream in self.ssl_streams for future use.
+        """
+        server_cert_path = self._get_cert_path(dest_replica.id, False)
+        client_cert_path = self._get_cert_path(self.client_id, True)
+        client_pk_path = self._get_private_key_path(self.client_id, True)
+        # Create an SSl context - enable CERT_REQUIRED and check_hostname = True
+        ssl_context = ssl.SSLContext(ssl.PROTOCOL_TLS_CLIENT)
+        # Verify server certificate using this trusted path
+        ssl_context.load_verify_locations(cafile=server_cert_path)
+        # Load my private key and certificate
+        ssl_context.load_cert_chain(client_cert_path, client_pk_path)
+        # Server hostname to be verified must be taken from create_tls_certs.sh
+        server_hostname = self.CERT_DOMAIN_FORMAT % dest_replica.id
+        while True:
+            try:
+                # Open TCP stream and connect to server
+                tcp_stream = await trio.open_tcp_stream(str(dest_replica.ip), int(dest_replica.port))
+                # Wrap this stream with SSL stream, pass server_hostname to be verified
+                ssl_stream = trio.SSLStream(tcp_stream, ssl_context, server_hostname=server_hostname, https_compatible=False)
+                # Wait for handshake to finish (we want to be on the safe side - after this we are sure 
+                # connection is open)
+                await ssl_stream.do_handshake()
+                # Success! keep stream in dictionary and break out
+                self.ssl_streams[(dest_replica.ip, dest_replica.port)] = ssl_stream
+                return
+            except (OSError, trio.BrokenResourceError):
+                await trio.sleep(0.25)
+                pass
+            continue
+
+    async def on_stopped_replicas(self, replica_ids):
+        """ BFT network inform client when a group of replicas are stopped. We relay on caller cancel scope """
+        for id in replica_ids:
+            replica = self.replicas[id]
+            addr = (replica.ip, replica.port)
+            if addr in self.ssl_streams:
+                if addr in self.ssl_streams:
+                    stream = self.ssl_streams[addr]
+                    del self.ssl_streams[addr]
+                    await stream.aclose()
+
+    async def _send_data(self, data, dest_replica):
+        """ Try to send data to dest_replica. On exception - close the SSL stream. """
+        dest_addr = (dest_replica.ip, dest_replica.port)
+        if dest_addr not in self.ssl_streams.keys():
+            # dest_replica is not connected (crushed? isolated?) 
```

Comment:
This is a perfectly normal scenario right? The replica can simply be down or partitioned as you indicate, and the client can't connect. But the higher level software doesn't know this and attempts to send, and so you simply return here. I'm just trying to clarify that we have the same understanding here. 

Additionally, when a partition heals, or a node restarts, we should reconnect from the client. In our C++ client we continuously try to reconnect in the background. We need some strategy for reconnecting.

Reply:
@andrewjstone (adding @teoparvanov here)- I'll try to shortly make things clearer here:
Yes - It's a normal scenario I agree.
At the beginning of the test, always before trying to send, when the replica is started, all clients connect to it. clients are always starting at the beginning (I'm not dealing with external clients for now). So if a replica is isolated or crushed the TCP socket stays in a special state (TIME_WAIT) and not yet closed. This is a 240-sec timeout, so no need to reconnect. If replica returns before that timeout, send will succeed since the socket is still open on both sides of no one explicitly closed it in the application layer. Every time when starting and stopping a replica, all clients are connected (TCP and SSL. handshake) and disconnected respectively.
My first implementation did exactly what you are suggesting - it did a lazy connection whenever trying to send data and connections is closed. This causes many complications, especially with the network partition manager, as the ports given to the client are unknown ports, and every time it happens, the connectivity port matrix in advisory needs to be refreshed. In addition, the client is unaware of who is healed, so it retries to reconnect all the time to all disconnected servers, and that causes a "major shake" in all tests - as it slows down all our estimated timeouts and multiple sends (tracker), and fails many of the tests.

As of now, In all our tests we get isolation/disconnection by:
1) Stopping and starting later some replica.
2) Isolating a replica for less than 4 minutes.
So all our tests pass.

If we want to isolate a replica for more than 4 minutes - it will require adding re-connect. But this is more complicated, especially to write notifications between the network partition manager and bft_client.

To summarize - I am aware of this, but this should be part of the 2nd stage impl., and not 1st stage (we should add a task for it already now)